### ISSUE TITLE
Handle the hostname header for the SSL verification

### ISSUE THREAD

**Issue by jhoukem** [Kong/unirest-java#368]
**Describe the bug**
I have a case where the computer on which my program run does not have a dns server. As a work around I use my server ip address to connect to my api. This lead to an SSL certificate error since the ip address  do not match the certificate hostname:  
` PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`  
On Postman I can reproduce the same issue but when I manually add the Host param in the header and make it match my server hostname it work properly.
When I do the same on Unirest the ip address is still used to check the ssl certificate.

**To Reproduce**
Make an https request to a server using its real ip in the url.

**Expected behavior**
When using an ip address in the url and the server hostname in the header under the "Host" var, the ssl verification should succeed.

If this is not a bug how would you advise me to proceed ?

**Comment by ryber** [comment]
I'll take a look but I'm not sure there is much that can be done here as the SSL validation is done down in Java and Unirest doesn't have a ton of control over it. It feels like what you are asking to do might be exploitable. In the mean time, you can disable SSL for local non-production purposes with ```Unirest.config().verifySsl(false);```

**Comment by jhoukem** [comment]
Thank you. From what I know it is possible to do so with apache if we use a specific HttpHost constructor: `HttpHost(final InetAddress address, final String hostname, final int port, final String scheme)` but in the unirest code running the one used is `HttpHost(final String hostname, final int port, final String scheme)` it is created in the CloseableHttpClient class in the `determineTarget` method.

From my understandung there is not much we can do about that but maybe I'm wrong...

**Comment by ryber** [comment]
Let me better understand what you are trying to do. When you say

> As a work around I use my server ip address to connect to my api

Is this all local, as in, you have this application, which has a built in certificate running locally on your computer, but because it's local you are connecting to your IP address rather than a hostname? 

and in addition, these certificates are self-signed?

I'm not sure you need to build an entire CloseableHttpClient as both the SSlContext and HostNameVerifier are exposed in the config:

```
SSLContext context = SSLContexts.custom()
                .loadTrustMaterial(null, new TrustSelfSignedStrategy())
                .build();
        HostnameVerifier verifier = (hostname, session) -> hostname.equals("www.zombo.com");

        Unirest.config().sslContext(context).hostnameVerifier(verifier);
```

Now, I wouldn't necessarily say this is something that should be in production code, but I also wouldn't host a production website with a self signed cert. For developer purposes its ok

**Comment by jhoukem** [comment]
>Is this all local, as in, you have this application, which has a built in certificate running locally on your computer, but because it's local you are connecting to your IP address rather than a hostname?

No the certificate is a basic Lets'encrypt one.

I use my ip address for the request  because the machine on which my program run on has no dns configured (my server hostname cannot be resolved).

This allow me to reach my webservice without having a dns (since it basically skip the resolution process by using the raw ip) but the counterpart is that now the SSL verification fails because it check for the hostname which is my raw ip address in this case (and the certificate specify the webservice hostname).

To try to fix this issue I added the header "host" to my request with my webservice hostname hoping that it will be used to check the SSL certificate (as it is the case in postman) but it turned out to be completely ignored by the implementation.

Furthermore when looking at the apache source code I see that this constructor`HttpHost(final InetAddress address, final String hostname, final int port, final String scheme)` would allow me to do what I want which is: to use my raw ip for the connection (InetAdress address) and use my hostname for the ssl verification (String hostname).

**Comment by ryber** [comment]
Just to follow up. I also cannot get Postman to do this swap on it's own. I get an error:

![postman](https://user-images.githubusercontent.com/145359/93790077-3b164d80-fbf8-11ea-8fc1-638ed70003aa.png)

**Comment by ryber** [comment]
Just for documentation curl doesn't support this:

```
curl -i -H "Host: sha512.badssl.com" https://104.154.89.105/
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

**Comment by jhoukem** [comment]
Yes indeed it seems that it is not supported by most clients out there. FYI I already tried with the apache client directly and it does work.
Your commit seems do to exactly what I need! It is awesome! Thank you for your time.
What is the next step now? Will it be merged or should I build it on my own?

**Comment by ryber** [comment]
I had to mull on it a bit. It's not exactly expected behavior, but I've also always seen postman as a closer relative to Unirest than Apache. I'll go ahead and merge it with some additional edge testing and get a release out this week.

**Comment by jhoukem** [comment]
Awesome thank you for that! 👍

**Comment by ryber** [comment]
released in 3.11.00
